### ISSUE TITLE
Figure out how to manage multiple themes

### ISSUE THREAD

**Issue by craigcook** [mozilla/protocol#442]
We knew this day would come...

As we're starting to integrate the Firefox brand and reorganize our approach to theming, we're faced with essentially three distinct themes for every component: a base theme, a Mozilla brand theme, and a Firefox brand theme. We're also facing theme variants, with light and dark options, so that can complicate things even further. There's even potential for incorporating more brand themes in the future (Pocket maybe?). We should take a moment to wrap our heads around this and plot a course of action with an eye toward maintenance and scalability.

One thing I've been pondering is how to handle this at the token level... do we want separate token libraries for each brand, or a single token library with everything in it? I can see plusses and minuses to either approach. 

Assuming a single-library scenario, we'd need to give tokens distinct names to avoid conflicts, especially colors because Mozilla's "red" is different from Firefox's "red". But it would simplify other things and there will be some tokens shared by both brands, like spacing and breakpoints. Another downside would be the possibility of bad token combinations, like mixing Mozilla colors with Firefox colors on the same page or within the same component. But perhaps we just need to trust ourselves not to do that, and namespacing brand-specific tokens would help there.

In a multi-library scenario there would be a lot of redundancy where themes might share the same token values (spacing and breakpoints again). There's also risk of breakage if something uses a token that only exists in one system, making it harder to switch themes without rebuilding the whole thing. But multiple libraries would forcibly prevent using the wrong brand's token because they wouldn't even be available, and it would prevent collisions of similar tokens.

I think I've talked myself into a single library and namespaced brand tokens but we should open the floor to other ideas and reach a consensus. 

And there's more to theme management than just tokens. We have to think about whether we even really need *every* component to exist in every theme. Are we just doubling our work? Before we go too far down any path, are we sure it's even a path worth taking? Maybe each brand should have a completeley separate system? Discuss!

**Comment by craigcook** [comment]
Further thoughts...

In working on the revised type scale for #410 I realized that the Firefox brand and Mozilla brand have different scales for headings. We size text with Sass mixins, and need to find a sensible way to maintain different type scales with minimal redundant code (I really don't want two whole sets of mixins). I think maybe we can rework those mixins to take some kind of theme variable.

**Comment by stephaniehobson** [comment]
It might be time for that fancy mixin you found that you need a table and a slide rule to figure out.

**Comment by alexgibson** [comment]
We should also probably spend some time thinking about how to handle supporting themes at the component level, in order to keep our CSS bundles as small as possible. We already publish our tokens as CSS custom variables, which have potential to make theming easier to manage. This may be worth some consideration?

**Comment by stephaniehobson** [comment]
So... no theming for nonsupporting browsers... intriguing.

**Comment by wboucherfischer** [comment]
What are the non-supported browsers?

**Comment by stephaniehobson** [comment]
Internet Explorer and old versions of other browsers: https://www.caniuse.com/#feat=css-variables

**Comment by stephaniehobson** [comment]
Some ideas:

## One theme per page all themes are dependent on a body class.

This is basically what we do now

```
.c-component {
  margin: 0 auto;

  .mzp-t-mozilla & {
     color: $mozilla-text-color;
  }

  .mzp-t-firefox & {
     color: $firefox-text-color;
  }
}
```

## All components switch to mixins with no output by default, then we pass the component theme:

```
@mixin c-component($theme) {
    @if($theme == 'mozilla') {
       $tokens = $mozilla-theme-tokens;  
   } @else {
       $tokens = $firefox-theme-tokens;  
   }

  .c-component {
     color: $tokens.textColor;
     font-size: $tokens.baseFontSize;
  }
}
```
```
@import 'protocol/component';
@include c-component('firefox');
```

## CSS Custom Properties (aka Variables):

All theme related values are put into CPs, let the cascade handle the rest. (Non supporting browsers would get a default theme)

```
.mzp-t-mozilla {
    --bodyTextColor: black;
    --baseFontSize: 16px;
}

.mzp-t-firefox {
    --bodyTextColor: ink;
    --baseFontSize: 18px;
}

.c-component {
   color: var(--bodyTextColor);
   font-size: var(--baseFontSize);
}
```

**Comment by stephaniehobson** [comment]
Further thoughts:

- We should try to write some requirements for what we want the system to do (do we need to have multiple themes on the same page?)
- Are we working on firefox vs mozilla or does the system need to handle light and dark variations as well?
- Do we want the system to be capable of minimizing the outputted CSS, or can we add some other kind of technical solution to solve that (postCSS?)
- Are we asking for trouble trying to handle both the Firefox and Mozilla themes together?

**Comment by craigcook** [comment]
I think we would want to avoid mixing two brands on the same page, so declaring a per-page theme might be a good way to forcibly prevent it. Then again, maybe there's a valid case for mixed branding I haven't thought of and we'd rather do it per component... or at last have some option to override where needed.

Either way, I think we'd treat the base "unthemed" theme as universal and brand-neutral, so things like navbars and footers can combine with either brand and not look out of place.

I've also been thinking more about the idea of "child themes" as an extension of a parent brand theme. So a dark variant would be a child of the parent brand theme, not a full-blown independent theme of its own. A child theme would inherit from its parent theme, and just change a few variables. If we go down the path of different color schemes for different Firefox products, that could be a child theme as well. Like if we decide Send uses orange and pink where Monitor uses purple and yellow, they could both be Firefox branded with different key colors defined in the child theme.

CSS custom properties might be the best way to do this, which then raises the issue of offering a fallback for lesser browsers. The essentials of what makes a "brand theme" might just boil down to the font and color palette, and most other aspects of the design would be shared across brands (spacing, breakpoints, etc). The unbranded default just uses Inter as the global font and that would be fine. But colors get a little more complicated. We'd need some universal palette for that unbranded default, then a brand theme can modify the default palette, and a child theme can modify the brand palette. 

So we're really just defining a palette with a handful of variables, and we need to design that and lock those in. A background color, a foreground color, an accent color, etc. We could boil it down to a palette of five or six key colors and a few alternatives.

I have also wondered if we're asking for trouble with all this... it's tempting to just go all-in on the Firefox brand, which could simplify everything. Yet we have so many sites that need a design system but are NOT Firefox branded, and I don't want to have two completely separate design systems.

**Comment by wboucherfischer** [comment]
I think you should talk to Clio about this. Frankly I think a lot of things
branded Mozilla should be Firefox.

**Comment by alexgibson** [comment]
> CSS custom properties might be the best way to do this, which then raises the issue of offering a fallback for lesser browsers.

This may not be viable for bedrock depending on the level of support we wish to deliver to older browsers, but if we used Custom Properties carefully, it's entirely possible to not code any fallback at all, and still have older browsers degrade gracefully to default browser colors. It may not be ideal, but it's still perfectly readable & accessible. Also considering most of our pages are black text on a white background, this may not be so drastic either.

**Comment by johngruen** [comment]
agree with alex that unfortunately, i don't believe w can support css custom properties at the moment. I have written what i view as an extremely simple proposal for resolving this issue as #447... the gist of it is `mzp-t-brand-firefox` or `mzp-t-brand-mozilla` would set a default document palette, but that we could add optional override classes for different colorways...ie`mzp-t-colorway-vivid`

**Comment by craigcook** [comment]
Some additional notes and brainstorming here, linked for posterity: https://docs.google.com/document/d/1DXQ3ZVNyXTGt_869fjeEJpRyX4TyhfYnf_vw7McGdbE/

I think we've discussed it enough and have a handle on the approach we want to take, and we'll file a separate issue for the implementation work.

tl;dr: Rather than hang everything from classes we think it's sensible to declare some global theme variables similar to how we presently declare the font and image paths. Those variables can then be used in functions/mixins to output different CSS. A rough proof of concept here: https://codepen.io/stephaniehobson/pen/BXwoBK?editors=0100
